File: furious-scribe.py
import asyncore, socket, argparse, StringIO, logging, uuid


HTTP_VERBS = ['GET', 'HEAD', 'POST', 'PUT', 'DELETE', 'TRACE', 'OPTIONS', 'CONNECT', 'PATCH']


class ToServer(asyncore.dispatcher):
    def __init__(self, to_client_socket, logger):
        global config
        asyncore.dispatcher.__init__(self)
        self.logger = logger
        self.to_client_socket = to_client_socket
        self.create_socket(socket.AF_INET, socket.SOCK_STREAM)
        self.connect((config.host, config.port))
        self.buffer = ''

    def handle_connect(self):
        pass

    def handle_close(self):
        self.close()

    def handle_read(self):
        data = self.recv(8192)
        if data:
            buffer = StringIO.StringIO(data)
            response_line = buffer.readline()
            self.logger.info('Response: %s' % response_line.strip())
            self.to_client_socket.send(data)
        else:
            self.logger.info('Server closed connection')
            self.to_client_socket.close()

    def writable(self):
        return len(self.buffer) > 0

    def handle_write(self):
        sent = self.send(self.buffer)
        self.buffer = self.buffer[sent:]


class ClientHandler(asyncore.dispatcher_with_send):
    def __init__(self, socket, logger):
        asyncore.dispatcher_with_send.__init__(self, socket)
        self.to_server = ToServer(self, logger)
        self.logger = logger

    def handle_read(self):
        data = self.recv(8192)
        if data:
            buffer = StringIO.StringIO(data)
            request_line = buffer.readline().split(' ')
            if len(request_line) == 3 and request_line[0].upper() in HTTP_VERBS:
                verb, path, _ = request_line
                self.logger.info('Request: %s %s' % (verb, path))
            else:
                self.logger.info('Got non-http  message from client (%d bytes "%s"' %
                                 (len(data), str(data[:100]+'...' if len(data) > 100 else str(data))))
            self.to_server.send(data)
        else:
            self.logger.info('Client closed connection')
            self.to_server.close()


class ToClient(asyncore.dispatcher):
    def __init__(self,):
        global config
        asyncore.dispatcher.__init__(self)
        self.create_socket(socket.AF_INET, socket.SOCK_STREAM)
        self.set_reuse_addr()
        self.bind(('localhost', config.listenport))
        self.listen(5)

    def handle_accept(self):
        global config
        pair = self.accept()
        if pair is not None:
            sock, addr = pair
            logger = self.get_logger()
            logger.info('Incoming connection from %s' % repr(addr))
            ClientHandler(sock, logger)

    def get_logger(self):
        id = str(uuid.uuid4())
        logger = logging.getLogger(id)
        logger.setLevel(logging.INFO)
        fh = logging.FileHandler('%s-%s-%s.log' % (config.host, config.port, id))
        fh.setLevel(logging.INFO)
        formatter = logging.Formatter('%(asctime)s - %(message)s')
        fh.setFormatter(formatter)
        logger.addHandler(fh)
        return logger


def get_config():
    parser = argparse.ArgumentParser(description='Logging proxy for debugging services.')
    parser.add_argument('--host', type=str)
    parser.add_argument('--port', type=int)
    parser.add_argument('--listenport', type=int)
    return parser.parse_args()


def main():
    global config
    config = get_config()
    ToClient()
    asyncore.loop()

if __name__ == "__main__":
    main()

File: furious-scribe.js
var net = require('net');
var http = require('http');
var log_module = require('log');
var uuid = require('uuid');
var fs = require('fs');
var spawn = require('child_process').spawn;

var config = require('./config.json');


function getUntilReady(url, callback) {
    http.get(url,function (res) {
        if (res.statusCode == 200)
            callback();
        else {
            console.warn("Got " + res.statusCode + " from [" + url + "]");
            setTimeout(function () {
                getUntilReady(url, callback);
            }, config.testRetryFrequency);
        }
    }).on('error', function (e) {
            console.error("Got error waiting for [" + url + "]: " + e.message);
            setTimeout(function () {
                getUntilReady(url, callback);
            }, config.testRetryFrequency);
        });
}


function runProxy(proxy) {
    proxy.clientHandler = net.createServer(function (clientSocket) {
        var log = new log_module('info', fs.createWriteStream(proxy.serverHost+'.'+proxy.serverPort+'.'+uuid.v4()+'.log'));
        log.info('Received connection from '+clientSocket.remoteAddress+':'+clientSocket.remotePort);
        var serverSocket = null;
        var bufferedData = [];
        serverSocket = new net.Socket();
        serverSocket.connect(parseInt(proxy.serverPort), proxy.serverHost, function () {
            proxy.started = true;
            if (bufferedData.length > 0) {
                var msg = Buffer.concat(bufferedData);
                serverSocket.write(msg);
                console.info("Sent " + msg.length + " bytes of buffered data");
                bufferedData = [];
            }
            serverSocket.on("data", function (data) {
                var response = data.toString();
                if (response.substring(0, 5) == 'HTTP/') {
                    log.info('Response: '+response.split(/\r?\n/)[0].trim())
                }
                clientSocket.write(data);
                log.info('Passed '+data.length+' bytes of data from server to client');
            });
            serverSocket.on('end', function () {
                log.info("Server hung up");
                clientSocket.end();
            });
            serverSocket.on('disconnect', function () {
                log.info("Server lost");
                clientSocket.destroy();
            });
        });
        clientSocket.on('data', function (msg) {
            var lines = msg.toString().split(/\r?\n/);
            if (lines.length > 0) {
                var header = lines[0];
                var header_parts = header.split(/\s+/);
                var verb = header_parts[0].toUpperCase();
                if (header_parts.length == 3 &&
                    (verb == 'GET' ||
                     verb == 'HEAD' ||
                     verb == 'POST' ||
                     verb == 'PUT' ||
                     verb == 'DELETE' ||
                     verb == 'TRACE' ||
                     verb == 'CONNECT' ||
                     verb == 'OPTIONS')) {
                    log.info('Request: '+verb+' '+header_parts[1].trim());
                }
            }
            serverSocket.write(msg);
            log.info('Passed '+msg.length+' bytes of data from client to server');
        });
        clientSocket.on('end', function () {
            log.info("Client hung up");
            serverSocket.end();
        });
        clientSocket.on('disconnect', function () {
            log.info("Client lost");
            serverSocket.destroy();
        });
    });

    proxy.clientHandler.listen(proxy.clientPort);
}

function runProxies() {
    for (var i in config.proxies) {
        if (config.proxies.hasOwnProperty(i)) {
            var proxy = config.proxies[i];
            runProxy(proxy);
        }
    }
}

runProxies();